osc-server/src/test/java/org/osc/core/broker/service/tasks/conformance/openstack/securitygroup/CreatePortGroupTaskTest.java

```diff
@@ -0,0 +1,302 @@
+	}
+
+	@Test
+	public void testExecute_SGWithMultipleSGM_ExecutionFinishes() throws Exception {
+
+		// Arrange.
+		SecurityGroup sg = createSecurityGroup(1L, "tenantId", "tenantName", 1L, "sgName");
+		sg.addSecurityGroupMember(newSGMWithPort(VM.class, 1L));
+		sg.addSecurityGroupMember(newSGMWithPort(VM.class, 2L));
+
+		List<NetworkElement> neList = ostRegisterPorts(sg);
+
+		String domainId = UUID.randomUUID().toString();
+		registerDomain(domainId, sg);
+
+		NetworkElement ne = createNetworkElement();
+
+		SdnRedirectionApi redirectionApi = mockRegisterNetworkElement(ne);
+		registerNetworkRedirectionApi(redirectionApi, sg.getVirtualizationConnector());
+
+		CreatePortGroupTask task = new CreatePortGroupTask(sg);
+
+		// Act.
+		task.execute();
+
+		// Assert.
+		verify(this.em, Mockito.times(1)).merge(sg);
+		Assert.assertEquals(ne.getElementId(), sg.getNetworkElementId());
+		assertNetworkElementsParentIdWithDomainId(neList, domainId);
+	}
+
+	@Test
+	public void testExecute_WhenDomainIsNotFound_ThrowsException() throws Exception {
+
+		// Arrange.
+		SecurityGroup sg = createSecurityGroup(1L, "tenantId", "tenantName", 1L, "sgName");
+		sg.addSecurityGroupMember(newSGMWithPort(VM.class, 1L));
+
+		ostRegisterPorts(sg);
+
+		// domain id null
+		registerDomain(null, sg);
+
+		this.exception.expect(Exception.class);
+		this.exception
+				.expectMessage(String.format("A domain was not found for the tenant: '%s' and Security Group: '%s",
+						sg.getTenantName(), sg.getName()));
+
+		CreatePortGroupTask task = new CreatePortGroupTask(sg);
+
+		// Act.
+		task.execute();
+
+		// Assert.
+		verify(this.em, Mockito.never()).merge(any());
+	}
+
+	@Test
+	public void testExecute_WhenPortGpIsNotFound_ThrowsException() throws Exception {
+
+		// Arrange.
+		SecurityGroup sg = createSecurityGroup(1L, "tenantId", "tenantName", 1L, "sgName");
+		sg.addSecurityGroupMember(newSGMWithPort(VM.class, 1L));
+
+		ostRegisterPorts(sg);
+
+		registerDomain(UUID.randomUUID().toString(), sg);
+
+		// network element/portGp null
+		NetworkElement ne = null;
+		SdnRedirectionApi redirectionApi = mockRegisterNetworkElement(ne);
+		registerNetworkRedirectionApi(redirectionApi, sg.getVirtualizationConnector());
+
+		this.exception.expect(Exception.class);
+		this.exception.expectMessage(String.format("RegisterNetworkElement failed to return PortGroup"));
+
+		CreatePortGroupTask task = new CreatePortGroupTask(sg);
+
+		// Act.
+		task.execute();
+
+		// Assert.
+		verify(this.em, Mockito.never()).merge(any());
+	}
+
+	private SecurityGroupMember newSGMWithPort(Class<? extends OsProtectionEntity> entityType, Long sgmId) {
+		VMPort port = null;
+		OsProtectionEntity protectionEntity;
+
+		if (entityType == VM.class) {
+			protectionEntity = new VM("region", UUID.randomUUID().toString(), "name");
+			port = newVMPort((VM) protectionEntity);
+		} else if (entityType == Network.class) {
+			protectionEntity = new Network("region", UUID.randomUUID().toString(), "name");
```

Comment:
Do you actually need all these types or would the VM type suffice for the tests? Looking at the class under test it seems there is nothing specific to the entity type. If the VM type is enough just simply always return a member of VM type and simplify this method.